src/utils.cairo

```diff
@@ -1,6 +1,94 @@
-use core::sha256::{compute_sha256_byte_array, compute_sha256_u32_array};
+use core::fmt::{Display, Formatter, Error};
 use core::num::traits::{Zero, One, BitSize};
-use core::starknet::secp256_trait::Secp256PointTrait;
+use core::sha256::{compute_sha256_byte_array, compute_sha256_u32_array};
+use core::to_byte_array::AppendFormattedToByteArray;
+
+#[derive(Copy, Drop, Debug)]
+// pub is required here as we define Hash in utils and we need to import Hash in merkle_tree.cairo
+// for arguments to merkle_root function
+pub struct Hash {
+    value: [u32; 8]
+}
+
+#[generate_trait]
+pub impl HashImpl of HashTrait {
+    #[inline(always)]
+    fn to_hash(array: [u32; 8]) -> Hash {
+        Hash { value: array }
+    }
+}
+
+impl HashDisplay of Display<Hash> {
+    fn fmt(self: @Hash, ref f: Formatter) -> Result<(), Error> {
+        let hash: u256 = (*self).into();
+        let base: u256 = 16;
+
+        hash.append_formatted_to_byte_array(ref f.buffer, base.try_into().unwrap());
+        Result::Ok(())
+    }
+}
+
+pub impl U256IntoHash of Into<u256, Hash> {
+    fn into(self: u256) -> Hash {
+        let mut result: Array<u32> = array![];
+        let mut low: u128 = self.low;
+        let mut high: u128 = self.high;
+
+        let mut i = 0;
+        loop {
+            if i == 4 {
+                break;
+            }
+            result.append((low & 0xffffffff).try_into().unwrap());
+            low = shr(low, 32_u32);
+            i += 1;
+        };
+
+        let mut i = 0;
+        loop {
+            if i == 4 {
+                break;
+            }
+            result.append((high & 0xffffffff).try_into().unwrap());
+            high = shr(high, 32_u32);
+            i += 1;
+        };
+
+        Hash {
+            value: [
+                *result[7],
+                *result[6],
+                *result[5],
+                *result[4],
+                *result[3],
+                *result[2],
+                *result[1],
+                *result[0],
+            ]
+        }
+    }
+}
+
+pub impl HashIntoU256 of Into<Hash, u256> {
+    fn into(self: Hash) -> u256 {
+        let [a, b, c, d, e, f, g, h] = self.value;
+
+        let mut low: u128 = 0;
+        let mut high: u128 = 0;
+
+        low = low | (a.into());
```

Comment:
`|` can be replaced by a `+` here. It would be interesting to see if it makes any difference in terms of gas usage.

Reply:
yes, i did that because i thought i would be more efficient but i didnt benchmark.
let me do it quickly

Reply:
indeed you were right!

with `|`

```
tristan@Tristans-MBP raito % scarb cairo-run --available-gas 2000000
   Compiling raito v0.1.0 (/Users/tristan/Desktop/Dev/raito/Scarb.toml)
    Finished release target(s) in 3 seconds
     Running raito
a0000000a0000000a0000000a0000000a0000000a0000000a0000000a
Run completed successfully, returning []
Remaining gas: 537608
```

with + 

```
tristan@Tristans-MBP raito % scarb cairo-run --available-gas 2000000
   Compiling raito v0.1.0 (/Users/tristan/Desktop/Dev/raito/Scarb.toml)
    Finished release target(s) in 3 seconds
     Running raito
a0000000a0000000a0000000a0000000a0000000a0000000a0000000a
Run completed successfully, returning []
Remaining gas: 541400
````

small improvement but we take it.

Reply:
done